megatron/mpu/layers.py

```diff
@@ -377,6 +385,30 @@ def backward(ctx, grad_output):
 
         return grad_input, grad_weight, grad_bias, None, None, None, None
 
+def create_q_linear(weight, bias=None, has_fp16_weights=False, threshold=6.0, index=None):
+    """
+    weight: any kind of weight is fine. fp32, bf16, or fp16. We assume this is a CPU weight, to be converted to int8 upon sending to GPU.
+        From Tim Dettmers: "when cuda() or to(device) is called, Int8Param class should intercept, cast the CPU weight to fp16 and do the transformation to int8 and then cast it to device/cuda."
+    bias: the actual bias tensor. Can also be fp32, bf16, or f16 (optional)
+    Other arg explanations TBD
+    """
+    output_features, input_features = weight.shape
+    has_bias = bias is not None
+    q_linear = Linear8bitLt(input_features, output_features, bias=has_bias, has_fp16_weights=has_fp16_weights, threshold=threshold, index=index)
+    q_linear.weight = weight
+    return q_linear
+
+def post_hook(module, incompatible_keys):
+    # we assume that the weight has been put on CPU and we disabled the initialization
+    has_bias = module.bias is not None
+    if has_bias:
+        raise Exception("LLM int8 conversion currently does not support bias. Assuming Zucchini model.")
+    module.q_linear = create_q_linear(module.weight, bias=None, has_fp16_weights=False, threshold=6.0, index=None)
```

Comment:
This is fairly generic. If there's another int8 method that requires some kind of conversion from fp16=>int8, put your logic for that here.